test/common/upstream/cluster_manager_impl_test.cc

```diff
@@ -3134,6 +3134,33 @@ TEST_F(ClusterManagerInitHelperTest, StaticSdsInitialize) {
   cluster1.initialize_callback_();
 }
 
+// Verify that primary cluster can be updated in warming state.
+TEST_F(ClusterManagerInitHelperTest, TestUpdateWarming) {
+  InSequence s;
+
+  auto sds = std::make_unique<NiceMock<MockClusterMockPrioritySet>>();
+  ON_CALL(*sds, initializePhase()).WillByDefault(Return(Cluster::InitializePhase::Primary));
+  EXPECT_CALL(*sds, initialize(_));
+  init_helper_.addCluster(*sds);
+  init_helper_.onStaticLoadComplete();
+
+  NiceMock<MockClusterMockPrioritySet> updated_sds;
+  ON_CALL(updated_sds, initializePhase()).WillByDefault(Return(Cluster::InitializePhase::Primary));
+  EXPECT_CALL(updated_sds, initialize(_));
+  init_helper_.addCluster(updated_sds);
+
+  // The override cluster is added. Manually drop the previous cluster. In production flow this is
+  // achieved by ClusterManagerImpl.
+  sds.reset();
```

Comment:
Is it possible to add a cluster manager test of this case that uses "real" DNS clusters? I think the main sequence that can actually happen here is:
1) CDS adds DNS cluster
2) DNS cluster begins init
3) CDS updates DNS cluster?

I think you could also pretty easily test this in your integration test below by doing:
1) Have CDS deliver a DNS cluster configured with health checking.
2) DNS cluster won't initialize pending health checking.
3) Update DNS cluster

?

Reply:
Thank you for the DNS with health check! 

After a quick search I didn't find an existing example but this test case
`TEST_F(ClusterManagerImplTest, RemoveWarmingCluster)` seem putting a `defaultStaticCluster("fake_cluster")` in warm.

Writing a test

Reply:
A slight difference is that DNS cluster is always depending on resolver. 

I modified the cluster to STATIC type to make it immediate ready.

Reply:
Thanks this all looks good, but can you add a real integration test for this in ADS integration test? I think it's really easy based on what you already have. Just use a static cluster with health checking, it will be stuck initializing, then you can update it?

/wait

Reply:
I had the impression that static cluster with bad health check doesn't block initialization... 
Yeah, adding integrate test to confirm initialization and the behavior health check behavior. Will udpate

Reply:
Added.

Also confirmed that failed health check in static cluster also turn the warm cluster to ready.
And STRICT DNS resolve failure has the same scary side effect.

I end up not accept the tcp connection but not write back to hold the DNS request.